### PR TITLE
feat: update argo image to v2.12.9 and automate update process. Fixes #5232

### LINKED ISSUE #5232: Upgrade argo image to 2.12+
Part of https://github.com/kubeflow/pipelines/issues/4553

/cc @NikeNano

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Fixes #5232

Changes:
* partially automate argo update process
* update argo cli to latest too, because [the previous blocker](https://github.com/argoproj/argo-workflows/issues/2793
) was resolved
* update argo manifests for breaking changes
* move argo manifests to `third-party` folder
* presubmit test for marketplace manifest

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Bobgy** [comment]
/hold
/wip

I think I shouldn't add the NOTICES folder to KFP source repo. It's better provide a script that can deterministically generate them.

**Comment by Bobgy** [comment]
/test kubeflow-pipelines-sdk-python36
/test kubeflow-pipelines-sdk-python37
/test kubeflow-pipelines-sdk-python38
/test kubeflow-pipelines-sdk-python39
/test kubeflow-pipelines-tfx-python36

These tests also used argo cli.

**Comment by Bobgy** [comment]
/cancel wip
/cancel hold

This is ready for review.

**Comment by chensun** [comment]
/lgtm

Thanks!

**Comment by Bobgy** [comment]
Looks like I need to incorporate a few more changes in https://github.com/kubeflow/pipelines/pull/4693.

**Comment by Bobgy** [comment]
/hold
I'd suggest letting https://github.com/kubeflow/pipelines/pull/5256 merge earlier.

File: backend/Dockerfile [hunk 1]
```diff
@@ -15,9 +15,11 @@ COPY backend/requirements.txt .
 RUN python3 -m pip install -r requirements.txt --no-cache-dir
 
 # Downloading Argo CLI so that the samples are validated
-#ADD  https://github.com/argoproj/argo/releases/download/v2.7.5/argo-linux-amd64 /usr/local/bin/argo
-ADD  https://github.com/argoproj/argo/releases/download/v2.4.3/argo-linux-amd64 /usr/local/bin/argo
-RUN chmod +x /usr/local/bin/argo
+ENV ARGO_VERSION v2.12.9
```
Comment (Ark-kun) [inline backend/Dockerfile:18]: Does the CLI work offline again?

File: manifests/kustomize/env/aws/config [hunk 2]
```diff
@@ -1,6 +1,5 @@
 {
 namespace: $(kfp-namespace),
-executorImage: gcr.io/ml-pipeline/argoexec:v2.7.5-license-compliance,
```
Comment (NikeNano) [inline manifests/kustomize/env/aws/config:3]: Do we need to replace this or how does it work?

File: third_party/argo/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# Third Party dependency -- Argo
+
+This folder holds resources for KFP to redistribute <https://argoproj.github.io/projects/argo>
+container images.
+
+There's no code change. KFP only makes sure the images comply with licenses of all its dependencies
```
Comment (NikeNano) [inline third_party/argo/README.md:6]: It is a bit unclear to me what we mean with this:
```
There's no code change.
```

File: backend/Dockerfile (hunk 1 above)
Comment (NikeNano) [reply backend/Dockerfile:18]: Not sure that it is an issue depending on what we need to do offline,  https://github.com/argoproj/argo-workflows/issues/2793#issuecomment-772633106. : 
```
You can lint offline, as long as you do not use template references.
```

do we depend on [template references](https://github.com/argoproj/argo-workflows/blob/master/docs/workflow-templates.md) and do we use the argo cli for other use cases than validating(lint) workflows?

**Comment by Ark-kun** [comment]
/lgtm
/hold
Unhold as needed.

File: manifests/gcp_marketplace/test/snapshot-base.yaml [hunk 4]
```diff
@@ -1638,6 +1656,7 @@ kind: Service
 metadata:
   name: minio-service
   labels:
+    app: minio
     app.kubernetes.io/name: my-release
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply manifests/gcp_marketplace/test/snapshot-base.yaml:1660]: snapshots are generated, I'll update their source

File: manifests/kustomize/env/aws/config (hunk 2 above)
Comment (Bobgy) [reply manifests/kustomize/env/aws/config:3]: it's also specified in workflow controller argument.
and it seems that it has been removed from https://github.com/argoproj/argo-workflows/blob/v2.12.9/docs/workflow-controller-configmap.yaml

File: third_party/argo/README.md (hunk 3 above)
Comment (Bobgy) [reply third_party/argo/README.md:6]: Understood the confusion, I meant we are not modifying argo workflow code, the redistribution only adds license compliance

**Comment by Bobgy** [comment]
I'll merge when all tests pass. It's better to iterate quick because there were structural changes.

**Pull request merged**